src/Utilities.js

```diff
@@ -0,0 +1,123 @@
+import * as CESymbols from './CustomElementInternalSymbols';
+
+const reservedTagList = new Set([
+  'annotation-xml',
+  'color-profile',
+  'font-face',
+  'font-face-src',
+  'font-face-uri',
+  'font-face-format',
+  'font-face-name',
+  'missing-glyph',
+]);
+
+/**
+ * @param {string} localName
+ * @returns {boolean}
+ */
+export function isValidCustomElementName(localName) {
+  const reserved = reservedTagList.has(localName);
+  const validForm = /^[a-z][.0-9_a-z]*-[\-.0-9_a-z]*$/.test(localName);
+  return !reserved && validForm;
+}
+
+/**
+ * @private
+ * @param {!Node} node
+ */
+export function isConnected(node) {
+  // Use `Node#isConnected`, if defined.
+  const nativeValue = node.isConnected;
+  if (nativeValue !== undefined) {
+    return nativeValue;
+  }
+
+  while (node && !(node instanceof Document)) {
+    node = node.parentNode || (node.nodeType === Node.DOCUMENT_FRAGMENT_NODE ? node.host : undefined);
+  }
+  return node instanceof Document;
+}
+
+/**
+ * @param {!Node} root
+ * @param {!function(!Element)} callback
+ */
+export function walkDeepDescendantElements(root, callback) {
```

Comment:
I'm not sure the design of "capture tree being added/removed into flat list before running callbacks, then call callbacks on the captured list" that was added in this PR is a good tradeoff of performance for correctness, at least as implemented.

In the V1/current V1 polyfills, the MO record just gave the parent node that was added removed, then it was treewalked and callbacks called during the walk.  This meant that if a connected callback of a parent removed children, those children would not have their connected callbacks called (which is technically incorrect, although they would just be re-disconnected immediately after).

Pre-capturing the list (which is more expensive since you have the added cost of constructing the array then iterating it) seems halfsies -- in the above example it ensures that connected callback _is_ called, but it seems like it will come _after_ the disconnected callback is run, which would be _more_ incorrect since the callbacks come out of order.  To correct this it seems like you'd need an even more correct (and expensive) callback queue, to ensure the callbacks are run in the correct order.  Like we've discussed before I'm ok with exploring that, but in the name of performance and time, it seems like the previous behavior (less correct but not broken, and faster) is a better tradeoff for the MVP changes we need right now.